src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/ComActivator.cs

```diff
@@ -125,6 +138,80 @@ public static object GetClassFactoryForType(ComActivationContext cxt)
+            Type? regFuncAttrType = Type.GetType($"System.Runtime.InteropServices.{attributeName}, System.Runtime.InteropServices", throwOnError: false);
+            if (regFuncAttrType == null)
+            {
+                // If the COM registration attributes can't be found then it is not on the type.
+                return;
+            }
+
+            if (!Path.IsPathRooted(cxt.AssemblyPath))
+            {
+                throw new ArgumentException();
+            }
+
+            Type classType = FindClassType(cxt.ClassId, cxt.AssemblyPath, cxt.AssemblyName, cxt.TypeName);
+
+            // Retrieve all the methods.
+            MethodInfo[] methods = classType.GetMethods(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static);
+
+            // Go through all the methods and check for the custom attribute.
+            foreach (MethodInfo method in methods)
+            {
+                // Check to see if the method has the custom attribute.
+                if (method.GetCustomAttributes(regFuncAttrType!, inherit: true).Length == 0)
+                {
+                    continue;
+                }
+
+                // Check to see if the method is static before we call it.
+                if (!method.IsStatic)
+                {
+                    string msg = register ? SR.InvalidOperation_NonStaticComRegFunction : SR.InvalidOperation_NonStaticComUnRegFunction;
+                    throw new InvalidOperationException(SR.Format(msg));
+                }
+
+                // Finally validate signature
+                ParameterInfo[] methParams = method.GetParameters();
+                if (method.ReturnType != typeof(void)
+                    || methParams == null
+                    || methParams.Length != 1
+                    || (methParams[0].ParameterType != typeof(string) && methParams[0].ParameterType != typeof(Type)))
+                {
+                    string msg = register ? SR.InvalidOperation_InvalidComRegFunctionSig : SR.InvalidOperation_InvalidComUnRegFunctionSig;
+                    throw new InvalidOperationException(SR.Format(msg));
+                }
+
+                // The function is valid so set up the arguments to call it.
+                var objs = new object[1];
+                if (methParams[0].ParameterType == typeof(string))
+                {
+                    // We are dealing with the string overload of the function - provide the registry key - see comhost.dll implementation
+                    objs[0] = $"HKEY_LOCAL_MACHINE\\SOFTWARE\\Classes\\CLSID\\{cxt.ClassId.ToString("B")}";
+                }
+                else
+                {
+                    // We are dealing with the type overload of the function.
+                    objs[0] = classType;
+                }
+
+                // Invoke the COM register function.
+                method.Invoke(null, objs);
+
+                return;
```

Comment:
Is it intentional that we stop on the first one?

Reply:
Yep. We only register the first one. That is the previous behavior but if it did find another an exception was thrown. My thought was there is no reason to keep search if we already registered one. I can see the point in letting the user know, but at this point it has been long enough to know that there can be only one. More than happy to be explicit if desired though.

Reply:
Matching .NET Framework semantics.